lighthouse-core/computed/javascript-duplication.js

```diff
@@ -0,0 +1,103 @@
+ */
+'use strict';
+
+const makeComputedArtifact = require('./computed-artifact.js');
+const JsBundles = require('./js-bundles.js');
+
+class JavascriptDuplication {
+  /**
+   * @param {string} source
+   */
+  static _normalizeSource(source) {
+    // Trim trailing question mark - b/c webpack.
+    source = source.replace(/\?$/, '');
+
+    // Normalize paths for dependencies by only keeping everything after the last `node_modules`.
+    const lastNodeModulesIndex = source.lastIndexOf('node_modules');
+    if (lastNodeModulesIndex !== -1) {
+      source = source.substring(lastNodeModulesIndex);
+    }
+
+    return source;
+  }
+
+  /**
+   * @param {string} source
+   */
+  static _shouldIgnoreSource(source) {
+    // Ignore bundle overhead.
+    if (source.includes('webpack/bootstrap')) return true;
+    if (source.includes('(webpack)/buildin')) return true;
+
+    // Ignore shims.
+    if (source.includes('external ')) return true;
+
+    return false;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   */
+  static async compute_(artifacts, context) {
+    const bundles = await JsBundles.request(artifacts, context);
+
+    /**
+     * @typedef SourceData
+     * @property {string} source
+     * @property {number} size
+     */
+
+    /** @type {Map<LH.Artifacts.RawSourceMap, SourceData[]>} */
+    const sourceDatasMap = new Map();
+
+    // Determine size of each `sources` entry.
+    for (const {rawMap, sizes} of bundles) {
+      /** @type {SourceData[]} */
+      const sourceDatas = [];
+      sourceDatasMap.set(rawMap, sourceDatas);
+
+      for (let i = 0; i < rawMap.sources.length; i++) {
+        if (this._shouldIgnoreSource(rawMap.sources[i])) continue;
+
+        const fullSource = (rawMap.sourceRoot || '') + rawMap.sources[i];
+        const sourceSize = sizes.files[fullSource];
+        sourceDatas.push({
+          source: JavascriptDuplication._normalizeSource(rawMap.sources[i]),
+          size: sourceSize,
+        });
+      }
+    }
+
+    /** @type {Map<string, Array<{scriptUrl: string, size: number}>>} */
+    const sourceDataAggregated = new Map();
+    for (const {rawMap, script} of bundles) {
+      const sourceDatas = sourceDatasMap.get(rawMap);
+      if (!sourceDatas) continue;
+
+      for (const sourceData of sourceDatas) {
+        let data = sourceDataAggregated.get(sourceData.source);
+        if (!data) {
+          data = [];
+          sourceDataAggregated.set(sourceData.source, data);
+        }
+        data.push({
+          scriptUrl: script.src || '',
+          size: sourceData.size,
+        });
+      }
+    }
+
+    for (const [key, value] of sourceDataAggregated.entries()) {
```

Comment:
I still get spooked by concurrentmodification from java days but the JS iterator can handle deleting keys it's already seen right?

Reply:
that's a pain in java b/c of its concurrency model. here, we know it's synchronous so it's all good.

https://stackoverflow.com/questions/38319649/do-i-need-to-worry-about-concurrent-modification-in-node-js

Reply:
well it goes beyond the concurrency model, some data structure iterators are based on an `i++` like cursor so if you update the underlying data structure *while traversing it*, they will throw a concurrentmodification exception even if the entire thing is single-threaded and never has concurrency to worry about

Reply:
the SO question would obviously never be a problem in node :)

but I could easily write a stupid iterator in node that would be funky if you spliced the underlying array

Reply:
ah well my understanding is that b/c Java does have a more complex concurrency story, they don't bother supporting the trivial modification in an iterator so that you don't get nasty bugs when you introduce concurrency to your program. Seems like if they wanted to, Java iterators could support this (without using the iterator [directly](https://stackoverflow.com/a/18448703/2788187))|

i'm glad I left the Java world and entered the _much simpler_ JS world ....